src/java.base/share/classes/java/lang/constant/DynamicConstantDesc.java

```diff
@@ -357,12 +357,11 @@ private static ConstantDesc canonicalizeArrayVarHandle(DynamicConstantDesc<?> de
     @Override
     public final boolean equals(Object o) {
         if (this == o) return true;
-        if (!(o instanceof DynamicConstantDesc)) return false;
-        DynamicConstantDesc<?> desc = (DynamicConstantDesc<?>) o;
-        return Objects.equals(bootstrapMethod, desc.bootstrapMethod) &&
-               Arrays.equals(bootstrapArgs, desc.bootstrapArgs) &&
-               Objects.equals(constantName, desc.constantName) &&
-               Objects.equals(constantType, desc.constantType);
+        return (o instanceof DynamicConstantDesc desc)
```

Comment:
should be
  `(o instanceof DynamicConstantDesc<?> desc)`

Reply:
I noticed that too initially. However, `javac` does not seem to produce a "rawtypes" warning. Which makes sense, if you think about it.

Reply:
The problem is that `desc` is a raw type and raw types doesn't play well with the rest of the language (in particular with inference).

I think it's a bug in javac, it should not even raise a warning but an error.
But the spec is not fully clear.

The spec says that the narrow conversion should not be unchecked
https://docs.oracle.com/javase/specs/jls/se15/preview/specs/patterns-instanceof-jls.html#jls-14.30.2
it depends if you consider the raw conversion as unchecked or not.

Reply:
> I think it's a bug in javac, it should not even raise a warning but an error.
> But the spec is not fully clear.

If you think that it's a bug, consider providing feedback to authors of JEP 394:

> Other refinements may be incorporated based on further feedback.

I find it hard to believe that authors didn't consider generic use case (even though JEP 394 doesn't have examples of that).

Reply:
We have considered generics, that why parameterized generics with another type arguments are forbidden, but i think we have forgotten raw types.

Reply:
I don't think that cast from `Object` to a raw type is unchecked, so as error does not seem warranted to me.

However, I agree javac should produce the rawtype warning for rawtypes in pattern matching instanceof, because we are introducing a new variable (and casting). I've filled:
https://bugs.openjdk.java.net/browse/JDK-8263590

Note the non-pattern matching instanceof does not produce the rawtype warning, and I don't think it should produce it.

Reply:
yes,
javac should emit a warning in that case, that also the answer of Brian.

https://mail.openjdk.java.net/pipermail/amber-spec-experts/2021-March/002925.html

Reply:
Then we should use an unbounded wildcard here and in similar places to avoid "rawtype" warnings in the future.

Reply:
I've added the wildcard to the pattern variable check now. See b89027c